src/extensions/default/NavigationAndHistory/main.js

```diff
@@ -0,0 +1,402 @@
+    var HIDE_TIMEOUT_DELAY = 1500,
+        OPEN_FILE_DELAY    = 600;
+    
+    /*
+    * Contains list of most recently opened files and their last known cursor position
+    * @private
+    * @type {Array.<Object>}
+    */
+    var _mrofList = [],
+        $mrofContainer,
+        _activePaneId = null;
+    
+    /**
+     * Opens a full editor for the given context
+     * @private
+     * @param {Object.<path, paneId, cursor>} contextData - wrapper to provide the information required to open a full editor
+     * @return {$.Promise} - from the commandmanager 
+     */
+    function _openEditorForContext(contextData) {
+        return CommandManager.execute(Commands.FILE_OPEN, {fullPath: contextData.path,
+                                                    paneId: contextData.paneId }).done(function () {
+            EditorManager.getActiveEditor().setCursorPos(contextData.cursor);
+            EditorManager.getActiveEditor().centerOnCursor();
+        });
+    }
+    
+    /**
+     * Creates an entry for MROF list
+     * @private
+     * @param {String} path - full path of a doc
+     * @param {String} pane - the pane holding the editor for the doc
+     * @param {Object} cursorPos - current cursor position
+     * @return {Object} a frame containing file path, pane and last known cursor
+     */
+    function _makeMROFListEntry(path, pane, cursorPos) {
+        return {
+            file: path,
+            paneId: pane,
+            cursor: cursorPos
+        };
+    }
+    
+    /**
+     * Determines if a file is dirty
+     * @private
+     * @param {!File} file - file to test
+     * @return {boolean} true if the file is dirty, false otherwise
+     */
+    function _isOpenAndDirty(file) {
+        // working set item might never have been opened; if so, then it's definitely not dirty
+        var docIfOpen = DocumentManager.getOpenDocumentForPath(file.fullPath);
+        return (docIfOpen && docIfOpen.isDirty);
+    }
+    
+    /** Returns a 'context' object for getting/setting project-specific preferences */
+    function _getPrefsContext() {
+        var projectRoot = ProjectManager.getProjectRoot();
+        return { location : { scope: "user", layer: "project", layerID: projectRoot && projectRoot.fullPath } };
+    }
+    
+    
+    function _checkExt(entry, index) {
+        var deferred = new $.Deferred(),
+            fileEntry = FileSystem.getFileForPath(entry.file);
+
+        fileEntry.exists(function (err, exists) {
+            if (!err && exists) {
+                deferred.resolve();
+            } else {
+                _mrofList[index] = null;
+                deferred.reject();
+            }
+        });
+
+        return deferred.promise();
+    }
+    
+    /**
+     * Checks whether entries in MROF list actually exists in fileSystem to prevent access to deleted files 
+     * @private
+     */
+    function _syncWithFileSystem() {
```

Comment:
The filter call actually removes all null,undefined entries from the array. The Async call then ckecks for individual file entries which might have been deleted or removed from the actual file system and then marks those entries as null. 
Then we filter the array again to remove null entries.